contracts/MarginPool.sol

```diff
@@ -3,10 +3,135 @@
  */
 pragma solidity 0.6.10;
 
+import {ERC20Interface} from "./interfaces/ERC20Interface.sol";
+import {AddressBookInterface} from "./interfaces/AddressBookInterface.sol";
+import {SafeMath} from "./packages/oz/SafeMath.sol";
+import {WETH9} from "./packages/canonical-weth/WETH9.sol";
+
 /**
- *
+ * @author Opyn Team
+ * @title MarginPool
+ * @notice contract that hold all protocol funds
  */
-// solhint-disable-next-line no-empty-blocks
 contract MarginPool {
+    using SafeMath for uint256;
+
+    /// @notice scaling unit
+    uint256 public constant BASE_UNIT = 1e18;
+
+    /// @notice AddressBook module
+    address public addressBook;
+
+    /// @notice WETH token
+    WETH9 public WETH;
+
+    /**
+     * @notice contructor
+     * @param _addressBook adressbook module
+     */
+    constructor(address _addressBook) public {
+        require(_addressBook != address(0), "Invalid address book");
+
+        addressBook = _addressBook;
+
+        WETH = WETH9(payable(AddressBookInterface(addressBook).getWeth()));
+    }
+
+    /**
+     * @notice check if the sender is the Controller module
+     */
+    modifier onlyController() {
+        require(
+            msg.sender == AddressBookInterface(addressBook).getController(),
+            "MarginPool: Sender is not Controller"
+        );
+
+        _;
+    }
+
+    /**
+     * @notice fallback function to reveive ETH
+     */
+    receive() external payable {}
+
+    /**
+     * @notice transfers asset from user to pool
+     * @dev all tokens are scaled to have 1e18 precision in contracts,
+     *      so amounts are scaled down to native token decimals using _calcTransferAmount().
+     *      If _asset equal to WETH address, transfer WETH from Controller address to pool
+     * @param _asset address of asset to transfer
+     * @param _user address of user to transfer assets from
+     * @param _amount amount of token to transfer from _user, scaled to 1e18 of precision
+     * @return true if successful transfer
+     */
+    function transferToPool(
+        address _asset,
+        address _user,
+        uint256 _amount
+    ) external onlyController returns (bool) {
+        require(_amount > 0, "MarginPool: transferToPool amount is below 0");
+
+        // get asset decimals
+        uint8 assetDecimal = ERC20Interface(_asset).decimals();
+        // scale amount
+        uint256 val = _calcTransferAmount(_amount, assetDecimal);
+
+        bool success;
+
+        /// check if asset is WETH or not, if WETH transfer the val from Controller address
+        if (_asset == address(WETH)) {
+            // tranfer WETH from controller to pool
```

Comment:
where does wrapping of the ETH happen? In what contract?

Reply:
in the controller level, somewhere in the operate function, should be on of the first steps.

Reply:
Right now the flow looks like following:

* Transfer to pool: `Controller` call WETH.deposit, and `Pool` pull tokens from Controller
* Transfer to user: `Pool` calls WETH.withdraw, and transfer ETH to user

I don't see a clear reason why we need to use `WETH` instead of transferring ETH in this case, because both Controller and Pool are having checks for ETH asset, so we are not simplifying anything IMO. 

I imagined using WETH would let us remove ETH check in one of the contracts, right now it's just the same as using ETH natively

Reply:
We used WETH so we don't use address(0) for ETH, as there are so many checks that use `address(0)` in the MarginAccount...etc